src/backend/Pg.cpp

```
@@ -61,90 +61,291 @@ PgResult::msg() const
+        }
+
+        if (!sent)
+        {
+            std::stringstream ss;
+            ss << "Can't send query: " << PQerrorMessage(conn_.get());
+            throw std::runtime_error(ss.str());
+        }
+
+        auto& socket = getSocket();
+
+        flush(yield);
+
+        /* Only read response if query was submitted successfully.
+           Only a single response is expected, but the API requires
+           responses to be read until nullptr is returned.
+           It is possible for pending reads on the connection to interfere
+           with the current query. For simplicity, this implementation
+           only flushes pending writes and assumes there are no pending reads.
+           To avoid this, all pending reads from each query must be consumed,
+           and all connections with any type of error be severed. */
+        while (true)
+        {
+            if (PQisBusy(conn_.get()))
+                socket.async_wait(boost::asio::ip::tcp::socket::wait_read,
+                    yield);
+
+            if (!PQconsumeInput(conn_.get()))
             {
-                // PQexecParams can process only a single command.
-                ret.reset(PQexecParams(
-                    conn_.get(),
-                    command,
-                    nParams,
-                    nullptr,
-                    values,
-                    nullptr,
-                    nullptr,
-                    0));
+                std::stringstream ss;
+                ss << "query consume input error: "
+                   << PQerrorMessage(conn_.get());
+                throw std::runtime_error(ss.str());
             }
-            else
+
+            if (PQisBusy(conn_.get()))
+                continue;
+            
+            pg_result_type res{PQgetResult(conn_.get()),
+                               [](PGresult *result)
+                               { PQclear(result); }};
+
+            if (!res)
+                break;
+
```

What do you want to do about multiple results? One possibility is to check if there's already a result, and then throw:

            if (ret)
            {
                Throw<std::runtime_error>("multiple results returned");
            }

The way it stands now is that if there are multiple commands submitted in the command*, such as "SELECT * FROM foo; SELECT * FROM bar", then this will return the results from the 2nd query.

Not a big deal either way. One way forces a single SQL statement per command submissions, the other allows multiple but silently loses the response to all but the last.

---

So I think we want to allow multiple queries for setting up the database and creating tables, so we do want to allow this behavior.

---

In that case, if one errors out, we should break and return the error. Otherwise a final success message should imply that all the queries were successful.

---

That's fine.